### PR TITLE
[FLINK-35886] Leave a note for future Flink 2.0 upgrade

### PR DESCRIPTION
**Pull request by pnowojski**
Fixing FLIP-471 in the old source would require some extra work, that we hopefully can avoid by removing the old source before doing the upgrade.

### CODE REVIEW

File: flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 1]
```diff
@@ -435,6 +435,13 @@ private List<KafkaTopicPartitionState<T, KPH>> createPartitionStateHolders(
                                         kafkaHandle,
                                         deserializedWatermarkStrategy.createTimestampAssigner(
                                                 () -> consumerMetricGroup),
+                                        // When upgrading to Flink 2.0, context has to provide also
+                                        // the input activity clock. This is not trivial for the old
+                                        // sources. Ideally we should drop this old source before
+                                        // this connector is upgraded to Flink 2.0. Otherwise, we
+                                        // can avoid the compilation error without fixing the
+                                        // FLIP-471 bug, by returning SystemClock as the input
+                                        // activity clock. Note that would be incorrect thing to do.
```
Comment (AHeise) [inline flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java:444]:
```suggestion
                                        // activity clock. Note that would be incorrect thing to do.
```

Wouldn't that more or less just reproduce the status quo? So while incorrect, it wouldn't be a regression.

If so, I'd rephrase `by returning SystemClock, which would reproduce the pre-FLIP-471 behavior`.

**Review by AHeise — COMMENTED** [review]
Left a note.

File: flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 2]
```diff
@@ -435,6 +435,13 @@ private List<KafkaTopicPartitionState<T, KPH>> createPartitionStateHolders(
                                         kafkaHandle,
                                         deserializedWatermarkStrategy.createTimestampAssigner(
                                                 () -> consumerMetricGroup),
+                                        // When upgrading to Flink 2.0, context has to provide also
+                                        // the input activity clock. This is not trivial for the old
+                                        // sources. Ideally we should drop this old source before
+                                        // this connector is upgraded to Flink 2.0. Otherwise, we
+                                        // can avoid the compilation error without fixing the
+                                        // FLIP-471 bug, by returning SystemClock as the input
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java:443]: That was a bug that required interface change to fix (so that's why FLIP).

[the remark this answers is not in the document]
Comment (AHeise) [reply flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java:443]: Ah it was not really clear to me that you are referring to the bug FLIP-471 is fixing. Please rephrase.
Suggestion: `the bug addressed by FLIP-471`.

**Review by AHeise — APPROVED** [review]
LGTM.
